### ISSUE TITLE
Make predictions faster with keeping gradient information

### ISSUE THREAD

**Issue by takafusui** [cornellius-gp/gpytorch#1051]
Dear all,

thank you at first developing an excellent library for the Gaussian process.
I use `GPyTorch` to solve a Gaussian process dynamic programming problem, where I solve optimization problems on all training points.
I use `torch.autograd.grad` to provide the gradient, the Jacobian and the Hessian, all three are necessary for an NLP optimizer.
When I profile my code, I realize that the `likelihood(model())` part is a bottleneck, but since I need to keep gradient information, I want to avoid using the `torch.no_grad()` option.

I summarize the total CPU time to solve one optimization problem with several options below:
| Options | CPU time [sec] | Gradient |
|---------|----------------|----------|
|`with torch.no_grad(), gpytorch.settings.fast_pred_var()` | 0.961 | Incorrect |
|`with gpytorch.settings.fast_pred_var()` | 2.129 | Correct |
| No option | 3.960 | Correct |

For my project, the automatic gradient information greatly stabilizes an optimizer when solving optimization problems.

I appreciate any comments on this.

**Comment by jacobrgardner** [comment]
Hi @takafusui,

Have you tried running your code on a GPU? If you have enough data that things are feeling slow, you'll probably get a significant speed lift from using even a relatively weak GPU.

**Comment by takafusui** [comment]
Hi @jacobrgardner 

thank you for your comment. Currently, the size of my training dataset is around 1000, meaning that GPU might be an overkill option I am afraid.

**Comment by jacobrgardner** [comment]
@takafusui So with n=1000 it depends, but if a single iteration is taking multiple seconds I would bet on a GPU giving a solid speedup. The constant factors on GPU ops are pretty good nowadays.

With that being said, at n=1000, there's really nothing you should be doing that should be taking 2 seconds short of like training the entire model for 50 iterations, even on a CPU. Could you paste some example code, maybe with your data / problem abstracted out to just a self contained GP example that is that slow?

**Comment by takafusui** [comment]
Hi @jacobrgardner,

thank you for your comment and sorry for my late replay. In the end, I decided to distribute optimization problems via `mpi4py` on the cluster but without GPU. I know that this is not a solution to solve the above issue, but I will go in this direction.

Once I can prepare a snippet to replicate my issue, I will paste it here. Thank you again for your kind help.
